apiserver/authorizercontrollermodel_test.go

```diff
@@ -0,0 +1,193 @@
+			controllerTag: controllerTag,
+		},
+		fallThroughAuthorizer:  fallThrough,
+		ModelAuthorizationInfo: s.modelAuthInfo,
+	}
+
+	err := authorizer.Authorize(c.Context(), authInfo)
+	c.Check(err, tc.ErrorIsNil)
+	c.Check(fallThroughCalled, tc.IsTrue)
+}
+
+// TestSuperUserControllerModelAllowed is testing that when the authorization
+// is for the controller model and the user has [permission.SuperuserAccess] on
+// the controller authorization is allowed.
+func (s *controllerModelAuthorizerSuite) TestSuperUserControllerModelAllowed(c *tc.C) {
+	defer s.setupMocks(c).Finish()
+
+	controllerUUID := tc.Must(c, uuid.NewUUID)
+	controllerTag := names.NewControllerTag(controllerUUID.String())
+	userTag := tc.Must1(c, names.ParseUserTag, "user-fred")
+
+	s.modelAuthInfo.EXPECT().IsAuthorizationForControllerModel(gomock.Any()).
+		Return(true)
+	s.permissionDelegator.EXPECT().SubjectPermissions(
+		gomock.Any(),
+		"fred",
+		corepermission.ID{
+			ObjectType: corepermission.Controller,
+			Key:        controllerUUID.String(),
+		},
+	).Return(corepermission.SuperuserAccess, nil)
+
+	var fallThrough authentication.AuthorizerFunc = func(
+		context.Context, authentication.AuthInfo,
+	) error {
+		c.Fatal("fall through authorizer should never have been called")
+		return nil
+	}
+
+	authInfo := authentication.AuthInfo{
+		Delegator: s.permissionDelegator,
+		Tag:       userTag,
+	}
+
+	authorizer := controllerModelPermissionAuthorizer{
+		controllerAdminAuthorizer: controllerAdminAuthorizer{
+			controllerTag: controllerTag,
+		},
+		fallThroughAuthorizer:  fallThrough,
+		ModelAuthorizationInfo: s.modelAuthInfo,
+	}
+
+	err := authorizer.Authorize(c.Context(), authInfo)
+	c.Check(err, tc.ErrorIsNil)
+}
+
+// TestNonSuperUserPermissionNotAllowed is testing that for any other permission
+// set in the controller that is not [permission.SuperuserAccess] the user is
+// not authorized.
+func (s *controllerModelAuthorizerSuite) TestNonSuperUserPermissionNotAllowed(c *tc.C) {
+	ctrl := s.setupMocks(c)
+	defer ctrl.Finish()
+
+	controllerUUID := tc.Must(c, uuid.NewUUID)
+	controllerTag := names.NewControllerTag(controllerUUID.String())
+	userTag := tc.Must1(c, names.ParseUserTag, "user-fred")
+
+	s.modelAuthInfo.EXPECT().IsAuthorizationForControllerModel(gomock.Any()).
+		Return(true).AnyTimes()
+	var fallThrough authentication.AuthorizerFunc = func(
+		context.Context, authentication.AuthInfo,
+	) error {
+		c.Fatal("fall through authorizer should never have been called")
+		return nil
+	}
+
+	notAllowedPermissions := []corepermission.Access{
+		corepermission.NoAccess,
+		corepermission.ReadAccess,
+		corepermission.WriteAccess,
+		corepermission.AdminAccess,
+		corepermission.LoginAccess,
+		corepermission.AddModelAccess,
+		corepermission.ConsumeAccess,
+	}
+
+	// Test that each not allowed permission does not authorize for the
+	// controller model.
+	for _, perm := range notAllowedPermissions {
+		c.Run(perm.String(), func(t *testing.T) {
```

Comment:
I've started just shadowing `c` by calling the `*testing.T` argument `c`.
Since below, you use the parent test's `c`.

Reply:
Thanks for finding that.